src/sentry/auth/password_validation.py

```diff
@@ -94,3 +101,47 @@ def get_help_text(self):
             "Your password must contain no more than %(max_length)d characters.",
             self.max_length,
         ) % {"max_length": self.max_length}
+
+
+class PwnedPasswordsValidator:
+    """
+    Validate whether a password has previously appeared in a data breach.
+    """
+
+    def __init__(self, threshold=1, timeout=0.200):
+        self.threshold = threshold
+        self.timeout = timeout
+
+    def validate(self, password, user=None):
+        digest = sha1(password.encode("utf-8")).hexdigest().upper()
+        prefix = digest[:5]
+        suffix = digest[5:]
+
+        url = f"https://api.pwnedpasswords.com/range/{prefix}"
+        headers = {
+            "User-Agent": "Sentry @ {}".format(options.get("system.url-prefix")),
+        }
+
+        try:
+            r = requests.get(url, headers=headers, timeout=self.timeout)
+        except Exception as e:
+            logger.warning(
+                "Unable to fetch PwnedPasswords API",
+                extra={
+                    "exception": str(e),
+                    "prefix": prefix,
+                },
+            )
+            return
+
+        for line in r.text.split("\n"):
+            if ":" not in line:
+                continue
+
+            breached_suffix, occurrences = line.rstrip().split(":")
+            if breached_suffix == suffix:
```

Comment:
I'm not certain if we _really_ need to do it here since they aren't full hashes, but should we use Django's `constant_time_compare` to avoid potential timing attacks?

Reply:
Nah, I think we're fine with normal comparison. This comparison happens by normal user's request when they try to register or to change password, not by attacker.